core/src/main/java/org/neo4j/graphalgo/core/neo4jview/GraphView.java

```diff
@@ -10,98 +10,89 @@
+                            consumer.accept(nodeId, toMappedNodeId(item.otherNode(originalNodeId)), item.id());
+                        });
                     }
-                }
+                });
             }
-        }
-    }*/
-
-    @Override
-    public void forEachRelation(int nodeId, Direction direction, RelationConsumer consumer) {
-        final long originalNodeId = toOriginalNodeId(nodeId);
-        try (Cursor<NodeItem> nodeItemCursor = read.nodeCursor(originalNodeId)) {
-
-            nodeItemCursor.forAll(nodeItem -> {
-                try (Cursor<RelationshipItem> relationships = nodeItem.relationships(mediate(direction))) {
-                    relationships.forAll(item -> {
-                        consumer.accept(nodeId, toMappedNodeId(item.otherNode(originalNodeId)), item.id());
-                    });
-                }
-            });
-        }
+        });
     }
 
     @Override
     public void forEachRelation(int nodeId, Direction direction, WeightedRelationConsumer consumer) {
-        try {
-            final RelationshipIterator iterator = read.nodeGetRelationships(nodeId, direction);
-            while (iterator.hasNext()) {
-                final long relationId = iterator.next();
-                final Cursor<RelationshipItem> relationshipItemCursor = read.relationshipCursor(relationId);
-                relationshipItemCursor.next();
-                final RelationshipItem item = relationshipItemCursor.get();
-                consumer.accept(
-                        nodeId,
-                        toMappedNodeId(item.otherNode(nodeId)),
-                        relationId,
-                        (double) read.relationshipGetProperty(relationId, propertyKey)
-                );
+        withinTransactionTyped(read -> {
+            try {
+                final RelationshipIterator iterator = read.nodeGetRelationships(nodeId, direction, relationTypeId);
+                while (iterator.hasNext()) {
+                    final long relationId = iterator.next();
+                    final Cursor<RelationshipItem> relationshipItemCursor = read.relationshipCursor(relationId);
+                    relationshipItemCursor.next();
+                    final RelationshipItem item = relationshipItemCursor.get();
+                    consumer.accept(
+                            nodeId,
+                            toMappedNodeId(item.otherNode(nodeId)),
+                            relationId,
+                            (double) read.relationshipGetProperty(relationId, propertyKey)
+                    );
+                }
+            } catch (EntityNotFoundException e) {
+                throw new RuntimeException(e);
             }
-        } catch (EntityNotFoundException e) {
-            throw new RuntimeException(e);
-        }
+            return null;
+        });
     }
 
     @Override
     public Iterator<WeightedRelationCursor> weightedRelationIterator(int nodeId, Direction direction) {
```

Comment:
document on all iterator methods that their resulting iterators can only be used safely the single, calling thread.